libs/string_util/include/hpx/string_util/split.hpp

```diff
@@ -0,0 +1,73 @@
+//  Copyright (c) 2020 ETH Zurich
+//  Copyright (c) 2002-2006 Pavol Droba
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_STRING_UTIL_SPLIT_HPP
+#define HPX_STRING_UTIL_SPLIT_HPP
+
+#include <algorithm>
+#include <cctype>
+#include <string>
+
+namespace hpx { namespace string_util {
+    namespace detail {
+        template <typename It, typename CharT, typename Traits,
+            typename Allocator>
+        std::basic_string<CharT, Traits, Allocator> substr(
+            std::basic_string<CharT, Traits, Allocator> const& s,
+            It const& first, It const& last)
+        {
+            std::size_t const pos = std::distance(std::begin(s), first);
+            std::size_t const count = std::distance(first, last);
+            return s.substr(pos, count);
+        }
+    }    // namespace detail
+
+    enum token_compress_mode_type
+    {
+        token_compress_off,
+        token_compress_on
+    };
+
+    template <typename Container, typename Predicate, typename CharT,
+        typename Traits, typename Allocator>
+    void split(Container& container,
+        std::basic_string<CharT, Traits, Allocator> const& str,
+        Predicate&& pred,
+        token_compress_mode_type compress_mode = token_compress_off)
+    {
+        container.clear();
+
+        auto token_begin = std::begin(str);
+        auto token_end = std::end(str);
+
+        do
+        {
+            token_end = std::find_if(token_begin, std::end(str), pred);
+
+            container.push_back(detail::substr(str, token_begin, token_end));
+
+            token_begin = token_end + 1;
```

Comment:
This might be dangerous if `std::find_if` above does not find any match.

Reply:
It is in fact safe because it doesn't get dereferenced in that case, but I'll try to rearrange it a bit to make it clear (or avoid updating `token_begin` in that case).

Reply:
Incrementing an end iterator is UB in any case, even if you don't dereference the result.

Reply:
I did not know that, very good to know! Thanks!